### PR TITLE
Dev/dpoole/tweak resource governance

### PR DESCRIPTION
**Pull request by dpoole73**
Tweak resource governance

1. increase the default memory cap (double it).  This is done based on experiments on sql SF nodes we saw memory related issues.  It is not expected to use more memory than before on normal machines
1. Make the CPU and Memory limits configurable from settings, allowing us to experiment without making further code changes
1. update vmwatch to the new version which supports additional parameeter to tell it how much memory it is allowed to use,  This is used to give the GC system more info and help keep the footprint down.

### CODE REVIEW

File: main/schema.go [hunk 1]
```diff
@@ -59,6 +59,16 @@ const (
           "type": "boolean",
           "default": false
         },
+        "memoryLimitInBytes": {
+          "description": "Optional - specifies the max memory that vmwatch can use",
+          "type": "integer",
+          "default": 80000000
+        },
+        "macCpuPercentage": {
```
Comment (frank-pang-msft) [inline main/schema.go:67]: maxCpuPercentage

File: main/vmWatch.go [hunk 2]
```diff
@@ -22,8 +22,8 @@ import (
 type VMWatchStatus string
 
 const (
-	MaxCpuQuota               = 1        // 1% cpu
-	MaxMemoryInBytes          = 40000000 // 40MB
+	DefaultMaxCpuPercentage   = 1        // 1% cpu
```
Comment (frank-pang-msft) [inline main/vmWatch.go:25]: I don't think we need these, since deserialization will follow the schema defaults you set in schema.go

File: main/vmWatch.go [hunk 3]
```diff
@@ -229,6 +229,28 @@ func setupVMWatchCommand(s *vmWatchSettings, hEnv HandlerEnvironment) (*exec.Cmd
 	args = append(args, "--heartbeat-file", GetVMWatchHeartbeatFilePath(hEnv))
 	args = append(args, "--execution-environment", GetExecutionEnvironment(hEnv))
 
+	// 0 is the default so allow that but any value below 30MB is not allowed
+	if s.MemoryLimitInBytes == 0 {
```
Comment (frank-pang-msft) [inline main/vmWatch.go:233]: can remove, see other comment

File: main/schema.go [hunk 4]
```diff
@@ -59,6 +59,16 @@ const (
...
+          "description": "Optional - specifies the max cpu that the vmwatch process is allowed to consume",
+          "type": "integer",
+          "default": 1
```
Comment (frank-pang-msft) [inline main/schema.go:70]: I think you can set max and min values here as well, see numberOfProbes as an example

**Review by frank-pang-msft — CHANGES_REQUESTED** [review]

File: main/vmWatch.go (hunk 3 above)
Comment (frank-pang-msft) [reply main/vmWatch.go:233]: This means almost all the other validation below might be removable as well

File: main/schema.go [hunk 5]
```diff
@@ -59,6 +59,16 @@ const (
...
+        },
```
Comment (frank-pang-msft) [inline main/schema.go:71]: can you add UTs to test memoryLimitInBytes and maxCpuPercentage setting scenarios (null, out of bounds, defaults, etc)

File: main/schema.go (hunk 4 above)
Comment (dpoole73) [reply main/schema.go:70]: cool, didn't know this schema stuff was so powerful!

File: main/schema.go (hunk 1 above)
Comment (dpoole73) [reply main/schema.go:67]: whoops!

File: main/vmWatch.go (hunk 2 above)
Comment (dpoole73) [reply main/vmWatch.go:25]: per discussion leaving the checks here

**Review by frank-pang-msft — APPROVED** [review]

**Review by zmyzheng — APPROVED** [review]
